src/ptf/testutils.py

```diff
@@ -1406,4 +1540,104 @@ def verify_packets_any(test, pkt, ports=[], device_number=0):
 
     test.assertTrue(received == True, "Did not receive pkt on any of ports %r for device %d" % (ports, device_number))
 
+def verify_any_packet_any_port(test, pkts=[], ports=[], device_number=0):
+    """
+    Check that _any_ of the packet is received on _any_ of the specified ports belonging to
+    the given device (default device_number is 0).
+
+    Also verifies that the packet is ot received on any other ports for this
+    device, and that no other packets are received on the device (unless --relax
+    is in effect).
+    """
+    received = False
+    index = 0
+    match_index = 0
+    (rcv_device, rcv_port, rcv_pkt, pkt_time) = test.dataplane.poll(
+        timeout=1,
+        filters=FILTERS
+    )
+    for port in ports:
+        if rcv_device != device_number:
+            continue
+        if port in ports:
+            for pkt in pkts:
+                logging.debug("Checking for pkt on device %d, port %d", device_number, port)
+                if str(rcv_pkt) == str(pkt) and rcv_port == port:
+                    received = True
+                    match_index = index
+        index += 1
+    verify_no_other_packets(test)
+
+    test.assertTrue(received == True, "Did not receive pkt on any of ports %r for device %d" % (ports, device_number))
+    return match_index
+
+def verify_packets_on_ports(test, pkts=[], ports=[], device_number=0):
+    """
+    Check that each packet is received on corresponding port in the port list belonging to
+    the given device (default device_number is 0).
+
+    Also verifies that the packet is ot received on any other ports for this
+    device, and that no other packets are received on the device (unless --relax
+    is in effect).
+    """
+    cnt = 0
+    pkt_cnt = 0
+    test.assertTrue(len(pkts) == len(ports), "packet list count does not match port list count")
+    for port in ports:
+        if port in ports:
+            pkt = pkts[cnt]
+            logging.debug("Checking for pkt on device %d, port %d", device_number, port)
+            #(rcv_device, rcv_port, rcv_pkt, pkt_time) = test.dataplane.poll(
+            (rcv_device, rcv_port, rcv_pkt, pkt_time) = dp_poll(
+                test, device_number,
+                port_number=port,
+                exp_pkt=pkt
+            )
+            if rcv_device != device_number:
+                continue
+
+            if rcv_pkt != None:
+                pkt_cnt += 1
+        cnt += 1
+
+    verify_no_other_packets(test)
+    test.assertTrue(pkt_cnt == len(ports), "Did not receive pkt on one of ports %r for device %d" % (ports, device_number))
+
+def verify_any_packet_on_ports_list(test, pkts=[], ports=[], device_number=0):
```

Comment:
I think this function is too much. I can live with the 2 above functions, but we're getting too specific here. This does not belong in the PTF core code.